### PR TITLE
Steam/Trace almost_equal method

### PR DESCRIPTION
**Pull request by d-chambers**
### What does this PR do?
This PR adds the `almost_equal` method to the `obspy.Stream` and `obspy.Trace` class definitions. These methods allow for a less strict comparison of waveform data in two ways:

1) Optionally, only the default parameters of stats objects are used in the comparisons. This is controlled by the `default_stats` parameter. In my experience the `processing` attribute can sometimes differ slightly between two steams that are otherwise identical, in which case it may still be desirable to consider them equal.

2) The numpy arrays of the traces are compared as "close" using the function `numpy.allclose`. This allows the user to specify relative and absolute tolerances for comparison using the `atol` and `rtol` parameters.  

### Why was it initiated?  Any relevant Issues?

This feature request was mentioned in #2269, and probably in other places before that. It will primarily be useful for obspy's test suite but may find usages in obspy-related projects as well. It will hopefully fix the windows issues the CI is running into.

It would be nice to implement an option that would raise an error if two streams/traces are not equal with a nice message explaining the differences, but that would require a little more energy/time than I currently have so it may be best left for a future PR. 

I am not 100% satisfied with the method name, so if anyone has other ideas please share.

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [x] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+ DOCS"
- [x] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)
- [x] All tests still pass.
- [x] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt` .

+DOCS

### CODE REVIEW

**Comment by d-chambers** [comment]
> Instead of having it as a method on the Stream/Trace classes I'd rather have it under some obspy.testing namespace

Sure no problem. Let's just make sure we are on the same page before I change anything. 
This is a bit more complicated than numpy because obspy has more than one data class. So should there be a `trace_almost_equal`, and `stream_almost_equal` function, or maybe just an `almost_equal` function and the types should be detected and the inputs dispatched accordingly? Of course if we ever want to extend this to `Catalog` or `Inventory` objects there may need to be different arguments and it may get rather jumbled. 

Also are you ok with the name? Maybe it should be `allclose` like numpy, or would that cause confusion?

File: obspy/core/trace.py [hunk 1]
```diff
@@ -2941,6 +2982,15 @@ def _data_sanity_checks(value):
         raise ValueError(msg)
 
 
+def _make_stats_dict(tr, default_stats):
+    """
+    Return a dict of stats from trace optionally including processing.
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply obspy/core/trace.py]: Do you think the `default_stats` param description is insufficient? 

        If True only compare the default stats on the traces, such as seed
        identification codes, start/end times, sampling_rates, etc. If
        False also compare extra stats attributes such as processing and
        format specific information.

**Comment by jkmacc-LANL** [comment]
Couldn’t you add some parameters to `Trace.__eq__` the same way you do to `Stream.__str__` for non-standard behavior?  Expert users could use the dunder method directly.

**Comment by d-chambers** [comment]
> Couldn’t you add some parameters to `Trace.__eq__`

We could, but that would make `Trace.__eq__` quite  a bit uglier and more complex. Also incurs the risk of introducing bugs to a commonly used function rather than just creating a new function. One the flip side, it would be one less function to remember. I don't really have a strong opinion on where this functionality lives but it is needed to (hopefully) fix the last CI failure that crops up on appveyor.

**Comment by megies** [comment]
Didn't look at the details, but the general idea is definitely something very nice to have for us @d-chambers  :rocket: :tada:

File: obspy/core/util/testing.py [hunk 2]
```diff
@@ -1050,5 +1050,167 @@ def setup_context_testcase(test_case, cm):
...
+    # First compare trace stats objects.
+    stats1 = _make_stats_dict(tr1, default_stats)
+    stats2 = _make_stats_dict(tr2, default_stats)
+    if stats1 != stats2:
+        return False
+    # Then  compare the array values
+    try:  # Use equal_nan if available
+        all_close = np.allclose(tr1.data, tr2.data, rtol=rtol,
+                                atol=atol, equal_nan=equal_nan)
+    except TypeError:
```
Comment (krischer) [inline obspy/core/util/testing.py]: We usually handle these issues by using https://github.com/obspy/obspy/blob/master/obspy/core/compatibility.py

An advantage of this is that all compatibility things are collected in one place and we could in theory periodically check and remove no longer needed ones.

File: obspy/core/util/testing.py [hunk 3]
```diff
@@ -1050,5 +1050,167 @@ def setup_context_testcase(test_case, cm):
...
+        # This happens on very old versions of numpy (< 1.7) which do not
+        # have isclose.
+        all_close = _all_close(tr1.data, tr2.data, rtol=rtol,
+                               atol=atol, equal_nan=equal_nan)
+    return all_close
+
+
+def _all_close(ar1, ar2, rtol, atol, equal_nan):
```
Comment (krischer) [inline obspy/core/util/testing.py]: Also move this to https://github.com/obspy/obspy/blob/master/obspy/core/compatibility.py.

**Comment by d-chambers** [comment]
>  How about calling the functions assert_traces_almost_equal() and assert_streams_almost_equal()

I like it. The parity with `numpy.testing.assert_almost_equal` is enough to convince me. It will also, as you say, make it easier to add better error messages in the future.

**Comment by d-chambers** [comment]
hmmm... this is frustrating. Seems like some unrelated failures are starting to crop up.

**Comment by d-chambers** [comment]
rebased to current master

**Comment by d-chambers** [comment]
> Is there anything you still want to add?

I think you are right about creating an obspy.testing module and I would like to add better error messages, but that seems like a 2.0 thing to me ;). Merging in, thanks for the review.

**Pull request merged**